File: Mvvm- MVI (Java)/app/src/main/java/com/mvvm/mvi/viewModels/LoginViewModel.java
package com.mvvm.mvi.viewModels;

import android.os.Handler;
import android.util.Log;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.mvvm.mvi.intents.login.LoginError;
import com.mvvm.mvi.intents.login.LoginNotInitiated;
import com.mvvm.mvi.intents.login.LoginProgressing;
import com.mvvm.mvi.intents.login.LoginStateIntent;
import com.mvvm.mvi.intents.login.LoginSuccess;
import com.mvvm.mvi.intents.userActions.UserAction;
import com.mvvm.mvi.intents.userActions.UserRestartClick;
import com.mvvm.mvi.intents.userActions.UserLoginClick;
import com.mvvm.mvi.models.LoginResponseModel;

public class LoginViewModel extends ViewModel {
    private MutableLiveData<LoginStateIntent> loginStateIntentLiveData = new MutableLiveData<>();

    public MutableLiveData<LoginStateIntent> getLoginStateIntentLiveData() {
        return loginStateIntentLiveData;
    }

    public void setLoginStateIntentLiveData(LoginStateIntent loginStateIntent) {
        this.loginStateIntentLiveData.setValue(loginStateIntent);
    }

    /**
     * Unidirectional flow with single entry point of userAction intents
     */
    public void processUserAction(UserAction userAction) {
        if (userAction instanceof UserLoginClick) {
            hitLoginApiGetToken(userAction);
        } else if (userAction instanceof UserRestartClick) {
            setLoginStateIntentLiveData(new LoginNotInitiated()); //login state changed to not initialised
        }
    }

    private void hitLoginApiGetToken(UserAction userAction) {
        setLoginStateIntentLiveData(new LoginProgressing()); //login state changed to processing

        UserLoginClick userLoginClick = (UserLoginClick) userAction;
        Log.i(this.getClass().getSimpleName(), userLoginClick.getUserName() + ":" + userLoginClick.getPassword());

        new Handler().postDelayed(new Runnable() { // Just replacement of Login API
            @Override
            public void run() {
                if (userLoginClick.getUserName().isEmpty() || userLoginClick.getPassword().isEmpty()) {
                    //login state changed to error
                    setLoginStateIntentLiveData(new LoginError("Error - Username and password can't be empty"));
                } else {
                    //login state changed to success result
                    LoginSuccess loginSuccess = new LoginSuccess(new LoginResponseModel("T_T_M_M_P_K_01", 5000));
                    setLoginStateIntentLiveData(loginSuccess);
                }
            }
        }, 4000);
    }
}

File:  MVI with ViewModel (Kotlin)/app/src/main/java/com/mvvm/mvi/viewModels/LoginViewModel.kt
package com.mvvm.mvi.viewModels

import android.os.Handler
import android.util.Log
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.mvvm.mvi.intents.login.LoginStateIntent
import com.mvvm.mvi.intents.login.LoginStateIntent.*
import com.mvvm.mvi.intents.userActions.UserAction
import com.mvvm.mvi.intents.userActions.UserAction.UserLoginClick
import com.mvvm.mvi.intents.userActions.UserAction.UserRestartClick
import com.mvvm.mvi.models.LoginResponseModel

class LoginViewModel : ViewModel() {
    val loginStateIntentLiveData = MutableLiveData<LoginStateIntent>()

    private fun setLoginStateIntentLiveData(loginStateIntent: LoginStateIntent) {
        loginStateIntentLiveData.value = loginStateIntent
    }

    /**
     * Unidirectional flow with single entry point of userAction intents
     */
    fun processUserAction(userAction: UserAction) {
        when(userAction) {
            is UserLoginClick -> hitLoginApiGetToken(userAction)
            is UserRestartClick -> setLoginStateIntentLiveData(LoginNotInitiated) //login state changed to not initialised
        }
    }

    private fun hitLoginApiGetToken(userAction: UserAction) {
        setLoginStateIntentLiveData(LoginProgressing) //login state changed to processing
        val userLoginClick = userAction as UserLoginClick
        Log.i(this.javaClass.simpleName, userLoginClick.userName + ":" + userLoginClick.password)

        Handler().postDelayed({
            if (userLoginClick.userName.isEmpty() || userLoginClick.password.isEmpty()) {
                //login state changed to error
                setLoginStateIntentLiveData(LoginError("Error - Username and password can't be empty"))
            } else {
                //login state changed to success result
                val loginSuccess = LoginSuccess(LoginResponseModel("T_T_M_M_P_K_01", 5000))
                setLoginStateIntentLiveData(loginSuccess)
            }
        }, 4000)
    }
}
